src/bmffimage.cpp

```diff
@@ -0,0 +1,502 @@
+        }
+        return result;
+    }
+
+    std::string BmffImage::boxName(uint32_t box)
+    {
+        char name[5];
+        std::memcpy(name, &box, 4);
+        name[4] = 0;
+        return std::string(name);
+    }
+
+    bool BmffImage::superBox(uint32_t box)
+    {
+        return box == TAG_moov || box == TAG_dinf || box == TAG_iprp || box == TAG_ipco || box == TAG_meta ||
+               box == TAG_iinf || box == TAG_iloc;
+    }
+
+    bool BmffImage::fullBox(uint32_t box)
+    {
+        return box == TAG_meta || box == TAG_iinf || box == TAG_iloc;
+    }
+
+    std::string BmffImage::mimeType() const
+    {
+        switch (fileType) {
+            case TAG_avif:
+                return "image/avif";
+            case TAG_heic:
+            case TAG_heif:
+                return "image/heif";
+            case TAG_crx:
+                return "image/x-canon-cr3";
+            default:
+                return "image/generic";
+        }
+    }
+
+    std::string BmffImage::uuidName(Exiv2::DataBuf& uuid)
+    {
+        const char* uuidCano = "\x85\xC0\xB6\x87\x82\xF\x11\xE0\x81\x11\xF4\xCE\x46\x2B\x6A\x48";
+        const char* uuidXmp = "\xBE\x7A\xCF\xCB\x97\xA9\x42\xE8\x9C\x71\x99\x94\x91\xE3\xAF\xAC";
+        const char* uuidCanp = "\xEA\xF4\x2B\x5E\x1C\x98\x4B\x88\xB9\xFB\xB7\xDC\x40\x6E\x4D\x16";
+        const char* result = std::memcmp(uuid.pData_, uuidCano, 16) == 0   ? "cano"
+                             : std::memcmp(uuid.pData_, uuidXmp, 16) == 0  ? "xmp"
+                             : std::memcmp(uuid.pData_, uuidCanp, 16) == 0 ? "canp"
+                                                                           : "";
+        return result;
+    }
+
+    long BmffImage::boxHandler(int depth /* =0 */)
+    {
+        long result = (long)io_->size();
+        long address = (long)io_->tell();
+        if (visits_.find(address) != visits_.end() || visits_.size() > visits_max_) {
+            throw Error(kerCorruptedMetadata);
+        }
+        visits_.insert(address);
+
+        BmffBoxHeader box = {0, 0};
+        if (io_->read((byte*)&box, sizeof(box)) != sizeof(box))
+            return result;
+
+        box.length = getLong((byte*)&box.length, bigEndian);
+        box.type = getLong((byte*)&box.type, bigEndian);
+#ifdef EXIV2_DEBUG_MESSAGES
+        bool bLF = true;
+        std::cerr << indent(depth) << "Exiv2::BmffImage::boxHandler: " << toAscii(box.type)
+                  << Internal::stringFormat(" %8ld->%u ", address, box.length);
+#endif
+        if (box.length == 1) {
+            DataBuf data(8);
+            io_->read(data.pData_, data.size_);
+            result = address + (long)getULongLong(data.pData_, littleEndian);
+            // sanity check
+            if (result < 0 || result > (long)io_->size()) {
+                result = (long)io_->size();
+                box.length = result - address;
+            }
+            std::cerr << Internal::stringFormat(" (%lu)", result);
```

Comment:
This printout should also be protected.  
How about moving the above printout after the `if (box.length == 1)` block to ensure the printout always prints the correct length?

Reply:
I much prefer to see both.  We only display the (long long length) for length=1.

For sure it should be guarded the EXIV2_DEBUG_MESSAGES